### PR TITLE
TensorFLow backend - Tensor

### PR DESCRIPTION
**Pull request by AGaliciaMartinez**
Implementation of DenseTensor class that wraps around  a `tensorflow.Tensor` object. Although all the methods have been implemented to have a working class, only `copy`, `to_array` and `__init__` are properly tested. The rest of the methods will be changed when added the specializations. Tests were adapted from QuTiP 5.0.

### CODE REVIEW

**Comment by AGaliciaMartinez** [comment]
Hey @jakelishman, @Ericgig and @quantshah. In the last meeting we had  a small discussion around how copy should be handled. I noticed thought that `tf.constant` does not perform a deep copy if the input is a `Tensor`. This means that the graph necessary for the auto diff will be preserved. The deep copy should be done by `tf.indentity`which does not preserve the graph of operations, it creates a new `Tensor` with the same values as the input `Tensor.` I was thinking that the default behaviour should be to preserve the graph with copy=False by default. What do you think?

**Comment by AGaliciaMartinez** [comment]
This is ready for review. I was wondering whether the name of the class should be DenseTensor, TfTensor or TensorflowTensor.

File: tests/core/data/test_tensordense.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        pytest.param(_valid_tensor(), {'shape': (1, 2, 3)}, ValueError,
+                     id="numpy-shape 3 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (-1, 1)}, ValueError,
+                     id="numpy-negative shape"),
+        pytest.param(_valid_tensor(), {'shape': (-4, -4)}, ValueError,
+                     id="numpy-both negative shape"),
+        pytest.param(_valid_tensor(), {'shape': (1213, 1217)}, ValueError,
+                     id="numpy-different shape"),
+    ])
+    def test_init_from_wrong_input(self, arg, kwargs, error):
```
Comment (Ericgig) [inline tests/core/data/test_tensordense.py:201]: This parametrisation is too long...
All error are value error, thus it it not needed.
You are also testing all tensor type ('arg') with each shape.
You could parametrize them independently. Also so `id`s are repeated.

File: src/qutip_tensorflow/core/data/dense_tensor.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            if len(shape) == 1:
+                shape = (shape[0], 1)
+
+        if not (
+            len(shape) == 2
+            and isinstance(shape[0], numbers.Integral)
+            and isinstance(shape[1], numbers.Integral)
+            and shape[0] > 0
+            and shape[1] > 0
+            and isinstance(shape, tuple)
```
Comment (Ericgig) [inline src/qutip_tensorflow/core/data/dense_tensor.py:52]: This test should be done first. If `shape` is an integer, `len(shape)` will raise an error and you will not get the desired error message.

File: tests/core/data/test_tensordense.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+    def test_copy_returns_a_correct_copy(self, data_tensor_dense):
+        """
+        Test that the copy() method produces an actual copy, and that the
+        result represents the same matrix.
+        """
+        original = data_tensor_dense
+        copy = data_tensor_dense.copy()
+        assert original is not copy
+        assert np.all(original._tf == copy._tf)
```
Comment (Ericgig) [inline tests/core/data/test_tensordense.py:223]: Is numpy the right tool to test tensor's equality?

**Review by Ericgig — COMMENTED** [review]
Have you tried to benchmark DenseTensor vs the direct tensor. I would be curious about timing for `transpose`, `adjoint`, `copy` and `trace` that you have here.

File: tests/core/data/test_tensordense.py (hunk 3 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_tensordense.py:223]: I have not found anything similar in TensorFlow.

File: tests/core/data/test_tensordense.py (hunk 1 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_tensordense.py:201]: I think now should be better.

**Comment by AGaliciaMartinez** [comment]
@Ericgig I decided to stop this PR and use PR #8 instead. PR #8 also contains code related to the "to" and "create" methods. I implemented there all the changes you suggested here. Even though I'm closing this PR, if there is any continuation on the conversation we were having It probably is better to have it here. If there is something new lets move it to PR #8. 

Regarding to the DenseTensor methods, I am planning to include a benchmark with the PR that will add the specialisation for these methods.

**Pull request closed without merging**